design-documents/redirect_domain_whitelist.md

```diff
@@ -0,0 +1,56 @@
+# Domain Whitelist for Frontend Redirects
+
+With the introduction of PWA as well as other server-client Magento solutions, the Magento host is acting as a headless server.
+This headless Magento server can serve multiple frontend applications, potentially on different domains.
+Currently there is no defined way that Magento knows which client domains it serves and are safe to redirect the user.
+This presents a problem for some PayPal payment solutions when trying to implement them through GraphQl.
+
+## Issue
+Several Paypal payment methods require some URL to redirect to once the payment has completed (e.g. [PayFlow](https://developer.paypal.com/docs/classic/payflow/integration-guide/#introducing-the-gateway-checkout-solutions)).
+Currently Magento uses the current host and some predefined controller that handles the redirect from PayPal.
+
+For the GraphQl implementation we want to support the scenario where a single "headless" Magento server will act as the API server for multiple frontend clients (possibly on different domains).
+In order to support this case, we need the ability for the frontend client to specify the URLs that PayPal will redirect the frontend to after the PayPal transaction.
+However, introducing the ability for the client to define the redirect URL opens a security concern.
+With an open redirect it could be possible for an attacker to manipulate this URL to send the user to a malicious website after payment.
+
+## Proposed Solution
+Currently, the Magento server has no way of knowing which domains all it's frontend clients are using.
+Therefore we are not able to screen the URLs provided by the client, to ensure they are safe.
+
+We propose introducing a new configuration that will contain a list of domains that are allowed in these redirect URLs.
+We can use this as a whitelist to validate redirect urls, significantly reducing the security concern.
+
+#### Configuration
+The domain whitelist configuration should live in `app/etc/env.php` for the following reasons:
+
+- This configuration file is already in use and well-known by implementors
+- env.php is a secure location that is only editable via file edit or cli command
+- The configuration could be set by setup scripts using `setup:config:set` cli command
+
+We will add a new option (`--domain_whitelist`) to the `setup:config:set` that accepts a comma separated list of domains.
```

Comment:
From example, it looks like subdomains should be specified directly. And what about wildcard for the domain names? Please, add this to the proposal.

Reply:
Clarified that wildcards should not be allowed